### PR TITLE
mds: fix stray CInodes' use-after-free bug when submit ELid entry

### PR DESCRIPTION
**Pull request by lxbsz**
When submitting a journal log entry it could start a new segment and it could advance the stray CInodes, which has been released just before it.

Fixes: commit 5a537476544("mds: introduce ELid event to create/close log")
Fixes: https://tracker.ceph.com/issues/62861

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -8013,8 +8019,6 @@ bool MDCache::shutdown_pass()
   // (only do this once!)
   if (!mds->mdlog->is_capped()) {
     dout(7) << "capping the mdlog" << dendl;
-    mds->mdlog->submit_entry(new ELid());
-    mds->mdlog->flush();
```
Comment (batrick) [inline src/mds/MDCache.cc:8017]: I think a better option here would be to skip `MDCache::advance_stray` call if the MDS is shutting down. What do you think?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mds/MDCache.cc (hunk 1 above)
Comment (lxbsz) [reply src/mds/MDCache.cc]: I am okay with this too. Fixed it. Thanks.

**Review by batrick — APPROVED** [review]

**Review by dparmar18 — APPROVED** [review]

File: src/mds/MDLog.cc [hunk 2]
```diff
@@ -286,7 +286,9 @@ LogSegment* MDLog::_start_new_segment(SegmentBoundary* sb)
   sb->set_seq(event_seq);
 
   // Adjust to next stray dir
-  mds->mdcache->advance_stray();
+  if (!mds->is_stopping() {
```
Comment (vshankar) [inline src/mds/MDLog.cc:289]: missing trailing `)`.

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/mds/MDLog.cc (hunk 2 above)
Comment (lxbsz) [reply src/mds/MDLog.cc]: Still in the local stash. Fixed it. Thanks @vshankar

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#26-Sep-2023

**Pull request merged**